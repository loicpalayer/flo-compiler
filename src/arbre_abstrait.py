import enum
import json
from typing import List, TypeAlias

JSON: TypeAlias = dict[str,
                       "JSON"] | list["JSON"] | str | int | float | bool | None


class Type(enum.Enum):
    INCONNU = enum.auto()
    ENTIER = enum.auto()
    BOOLEEN = enum.auto()
    FONCTION = enum.auto()
    AUTRE = enum.auto()


class AST:

    def to_json(self) -> JSON:
        print(self.__class__.__name__)
        raise NotImplementedError("to_json not implemented")

    def type(self) -> Type:
        return Type.INCONNU

    def __str__(self):
        return json.dumps(self.to_json(), indent=2)


class Programme(AST):

    def __init__(self, instructions: List[AST]):
        self.instructions = instructions

    def type(self) -> Type:
        return Type.AUTRE

    def to_json(self) -> JSON:
        return {"instructions": [i.to_json() for i in self.instructions]}

    def addInstruction(self, instruction: AST):
        self.instructions.append(instruction)


class Operation(AST):

    def __init__(self, op: str, exp1: AST, exp2: AST):
        self.lhs = exp1
        self.op = op
        self.rhs = exp2

    def type(self) -> Type:
        if self.lhs.type() == self.rhs.type():
            return self.lhs.type()
        return Type.INCONNU

    def to_json(self) -> JSON:
        return {
            "op": self.op,
            "lhs": self.lhs.to_json(),
            "rhs": self.rhs.to_json()
        }


class OperationUnaire(AST):

    def __init__(self, op: str, exp: AST):
        print("OperationUnaire", op, exp)
        self.op = op
        self.exp = exp

    def type(self) -> Type:
        return self.exp.type()

    def to_json(self) -> JSON:
        return {"op": self.op, "exp": self.exp.to_json()}


class Entier(AST):

    def __init__(self, valeur: int):
        self.valeur = valeur

    def type(self) -> Type:
        return Type.ENTIER

    def __eq__(self, rhs: object) -> bool:
        if isinstance(rhs, Entier):
            return self.valeur == rhs.valeur
        return self.valeur == rhs

    def to_json(self) -> JSON:
        return {"entier": self.valeur}


class Booleen(AST):

    def __init__(self, valeur):
        self.valeur = valeur

    def type(self) -> Type:
        return Type.BOOLEEN

    def __eq__(self, rhs: object) -> bool:
        if isinstance(rhs, Booleen):
            return self.valeur == rhs.valeur
        return self.valeur == rhs

    def to_json(self) -> JSON:
        return {"booleen": self.valeur}


class Identifiant(AST):

    def __init__(self, valeur: str):
        self.valeur = valeur

    def type(self) -> Type:
        return Type.AUTRE

    def __eq__(self, rhs: object) -> bool:
        if isinstance(rhs, Identifiant):
            return self.valeur == rhs.valeur
        return self.valeur == rhs

    def to_json(self) -> JSON:
        return {"identifiant": self.valeur}


class FunctionArgs(AST):

    def __init__(self, args: List[AST]):
        self.args = args

    def type(self) -> Type:
        return Type.AUTRE

    def to_json(self) -> JSON:
        return list(map(lambda x: x.to_json(), self.args))

    def __getitem__(self, i):
        return self.args[i]

    def append(self, arg: AST):
        self.args.append(arg)


class AppelFonction(AST):

    def __init__(self, name: Identifiant, args: FunctionArgs):
        self.name = name
        self.args = args

    def type(self) -> Type:
        return Type.INCONNU

    def to_json(self) -> JSON:
        return {"function_name": self.name.valeur, "args": self.args.to_json()}
