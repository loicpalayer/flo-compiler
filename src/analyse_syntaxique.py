import sys
from sly import Parser
from src.analyse_lexicale import FloLexer
import src.arbre_abstrait as arbre_abstrait


class FloParser(Parser):
    # On récupère la liste des lexèmes de l'analyse lexicale
    tokens = FloLexer.tokens

    precedence = (
        ('left', 'ET'),
        ('left', 'OU'),
        ('right', 'NON'),
        ('right', 'EGAL'),
        ('left', 'DIFFERENT', 'INFERIEUR_OU_EGAL', 'SUPERIEUR_OU_EGAL',
         'INFERIEUR', 'SUPERIEUR'),
        ('left', '+', '-'),
        ('left', '*', '/', '%'),
        ('nonassoc', 'UMOINS'),
    )

    # Règles gramaticales et actions associées

    @_('listeInstructions')
    def prog(self, p):
        return arbre_abstrait.Programme(p[0])

    @_('instruction')
    def listeInstructions(self, p):
        l = arbre_abstrait.ListeInstructions()
        l.instructions.append(p[0])
        return l

    @_('instruction listeInstructions')
    def listeInstructions(self, p):
        p[1].instructions.append(p[0])
        return p[1]

    @_('ECRIRE "(" expr ")" ";"')
    def instruction(self, p):
        return arbre_abstrait.Ecrire(p.expr)

    @_('bool')
    def expr(self, p):
        return p[0]

    @_('BOOLEEN')
    def bool(self, p):
        return arbre_abstrait.Booleen(p[0])

    @_('bool ET bool')
    def bool(self, p):
        return arbre_abstrait.Operation('et', p[0], p[2])

    @_('bool OU bool')
    def bool(self, p):
        return arbre_abstrait.Operation('ou', p[0], p[2])

    @_('NON bool')
    def bool(self, p):
        return arbre_abstrait.OperationUnaire('non', p[1])

    @_('exprArith')
    def bool(self, p):
        return p[0]

    @_('exprArith "+" exprArith', 'exprArith "-" exprArith',
       'exprArith "*" exprArith', 'exprArith "/" exprArith',
       'exprArith "%" exprArith', 'exprArith INFERIEUR exprArith',
       'exprArith SUPERIEUR exprArith', 'exprArith EGAL exprArith',
       'exprArith DIFFERENT exprArith',
       'exprArith INFERIEUR_OU_EGAL exprArith',
       'exprArith SUPERIEUR_OU_EGAL exprArith')
    def exprArith(self, p):
        return arbre_abstrait.Operation(p[1], p[0], p[2])

    @_('"-" exprArith %prec UMOINS')
    def exprArith(self, p):
        return arbre_abstrait.OperationUnaire('-', p[1])

    @_('"(" bool ")"')
    def exprArith(self, p):
        return p[1]

    @_('facteur')
    def exprArith(self, p):
        return p[0]

#    @_('IDENTIFIANT "(" function_arg ")"')
#    def facteur(self, p):
#        return arbre_abstrait.AppelFonction(p.IDENTIFIANT, p.function_arg)

    @_('ENTIER')
    def facteur(self, p):
        return arbre_abstrait.Entier(p.ENTIER)

    @_('IDENTIFIANT')
    def facteur(self, p):
        return arbre_abstrait.Identifiant(p.IDENTIFIANT)


#    @_('expr "," function_arg ')
#    def function_arg(self, p):
#        p.function_arg.append(p.expr)
#        return p.function_arg

#    @_('expr')
#    def function_arg(self, p):
#        return [p.expr]


def analyse_syntaxique(input):
    lexer = FloLexer()
    parser = FloParser()
    return parser.parse(lexer.tokenize(input)).to_json()
