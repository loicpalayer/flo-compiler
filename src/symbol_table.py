from typing import Dict, List

from src.arbre_abstrait import Declaration, Identifiant, Programme, Type, Function


class SymbolTable:

    def __init__(self):
        self._symbols: Dict[str, Function] = {}

    @staticmethod
    def from_program(program: Programme):
        symbol_table = SymbolTable()
        for instruction in program.instructions:
            if isinstance(instruction, Function):
                symbol_table.add(instruction)
        return symbol_table

    def check_presence(self, func: Identifiant):
        if not func in self:
            raise Exception(f"La fonction {func} n'est pas définie")

    def add(self, func: Function):
        if func in self:
            raise Exception(f"La fonction {func.name} est déjà définie")
        self._symbols[func.name.valeur] = func

    def type(self, name: Identifiant) -> Type:
        self.check_presence(name)
        return self._symbols[name.valeur].return_type

    def args(self, func: Identifiant) -> List[Declaration]:
        self.check_presence(func)
        return self._symbols[func.valeur].args

    def memory_size(self, func: Identifiant) -> int:
        self.check_presence(func)
        return 4 * len(self.args(func))

    def __contains__(self, func: Function | Identifiant):
        if isinstance(func, Identifiant):
            return func.valeur in self._symbols

        return func.name.valeur in self._symbols

    def __str__(self):
        return '\n'.join([str(symbol) for symbol in self._symbols.values()])