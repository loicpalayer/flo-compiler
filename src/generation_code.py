import sys
from typing import List, Optional
from src.analyse_syntaxique import analyse_syntaxique
import src.arbre_abstrait as arbre_abstrait
from src.symbol_table import SymbolTable

num_etiquette_courante = -1  #Permet de donner des noms différents à toutes les étiquettes (en les appelant e0, e1,e2,...)

afficher_table = False
afficher_nasm = True

current_function: Optional[arbre_abstrait.Function] = None


def check_type(expr: arbre_abstrait.AST, expected: arbre_abstrait.Type):
    """
    Fonction locale, permet de vérifier que le type d'une expression est bien celui attendu.
    """
    if expr.type() != expected:
        raise Exception(
            f"Erreur de type: {expr} est de type {expr.type()} au lieu de {expected}"
        )


def printifm(*args, **kwargs):
    """
    Un print qui ne fonctionne que si la variable afficher_table vaut Vrai.
    (permet de choisir si on affiche le code assembleur ou la table des symboles)
    """
    if afficher_nasm:
        print(*args, **kwargs)


def printift(*args, **kwargs):
    """
    Un print qui ne fonctionne que si la variable afficher_table vaut Vrai.
    (permet de choisir si on affiche le code assembleur ou la table des symboles)
    """
    if afficher_table:
        print(*args, **kwargs)


def nasm_comment(comment):
    """
    Fonction locale, permet d'afficher un commentaire dans le code nasm.
    """
    if comment != "":
        printifm(
            "\t\t ; " + comment
        )  #le point virgule indique le début d'un commentaire en nasm. Les tabulations sont là pour faire jolie.
    else:
        printifm("")


def nasm_instruction(opcode, op1="", op2="", op3="", comment=""):
    """
    Affiche une instruction nasm sur une ligne
    Par convention, les derniers opérandes sont nuls si l'opération a moins de 3 arguments.
    """
    if op2 == "":
        printifm("\t" + opcode + "\t" + op1 + "\t\t", end="")
    elif op3 == "":
        printifm("\t" + opcode + "\t" + op1 + ",\t" + op2 + "\t", end="")
    else:
        printifm("\t" + opcode + "\t" + op1 + ",\t" + op2 + ",\t" + op3,
                 end="")
    nasm_comment(comment)


def nasm_nouvelle_etiquette():
    """
    Retourne le nom d'une nouvelle étiquette
    """
    global num_etiquette_courante
    num_etiquette_courante += 1
    return "e" + str(num_etiquette_courante)


def gen_entrypoint(programme: arbre_abstrait.Programme):
    """
    Affiche le code nasm correspondant à tout un programme
    """
    printifm('%include\t"io.asm"')
    printifm('section\t.bss')
    printifm(
        'sinput:	resb	255	;reserve a 255 byte space in memory for the users input string'
    )
    printifm('v$a:	resd	1')
    printifm('section\t.text')
    printifm('global _start')

    symbol_table = SymbolTable.from_program(programme)
    printift(symbol_table)

    gen_programme(programme, symbol_table, is_main=True)

    nasm_instruction("mov", "eax", "1", "", "1 est le code de SYS_EXIT")
    nasm_instruction("int", "0x80", "", "", "exit")


def gen_programme(programme: arbre_abstrait.Programme,
                  symbol_table: SymbolTable,
                  is_main=False):

    funcs = [
        instruction for instruction in programme.instructions
        if isinstance(instruction, arbre_abstrait.Function)
    ]
    for func in funcs:
        gen_function(func, symbol_table)

    if is_main:
        printifm('_start:')

    instrs = [
        instruction for instruction in programme.instructions
        if not isinstance(instruction, arbre_abstrait.Function)
    ]
    gen_listeInstructions(instrs, symbol_table)


def gen_listeInstructions(instructions: List[arbre_abstrait.AST],
                          symbol_table: SymbolTable):
    """
    Affiche le code nasm correspondant à une suite d'instructions
    """
    for instruction in instructions:
        gen_instruction(instruction, symbol_table)


def gen_return(return_: arbre_abstrait.Return, symbol_table: SymbolTable):
    """
    Affiche le code nasm correspondant à un return. Met sa valeur dans eax avant de revenir
à l’appel de la fonction grâce à l’instruction nasm ret
    """

    if current_function is None:
        raise Exception("return en dehors d'une fonction")

    if current_function.return_type != return_.value.type():
        raise Exception("return de type différent de la fonction")

    gen_expression(return_.value, symbol_table)
    nasm_instruction("pop", "eax", "", "", "")
    nasm_instruction("ret", "", "", "", "retourne à l'appel de la fonction")


def gen_instruction(instruction: arbre_abstrait.AST,
                    symbol_table: SymbolTable):
    """
    Affiche le code nasm correspondant à une instruction
    """
    if isinstance(instruction, arbre_abstrait.Function):
        gen_function(instruction, symbol_table)
    elif isinstance(instruction, arbre_abstrait.Return):
        gen_return(instruction, symbol_table)
    else:
        gen_expression(instruction, symbol_table)


class Builtins:

    @staticmethod
    def ecrire(arg: arbre_abstrait.AST, symbol_table: SymbolTable):
        """
        Affiche le code nasm correspondant au fait d'envoyer la valeur entière d'une expression sur la sortie standard
        """
        # on calcule et empile la valeur d'expression
        gen_expression(arg, symbol_table)
        # on dépile la valeur d'expression sur eax
        nasm_instruction("pop", "eax", "", "", "")
        # on envoie la valeur d'eax sur la sortie standard
        nasm_instruction("call", "iprintLF", "", "", "")

    @staticmethod
    def lire(symbol_table: SymbolTable):
        """
        Affiche le code nasm correspondant au fait de lire un entier sur l'entrée standard et de le mettre dans une variable
        """
        nasm_instruction("mov", "eax", "sinput", "", "")
        nasm_instruction("call", "readline", "", "", "")
        nasm_instruction("call", "atoi", "", "", "")
        nasm_instruction("push", "eax", "", "", "")


def gen_appel_fonction(fonction: arbre_abstrait.AppelFonction,
                       symbol_table: SymbolTable):
    """
    Affiche le code nasm correspondant à l'appel d'une fonction
    """
    if fct := getattr(Builtins, fonction.name.valeur, None):
        fct(*fonction.args, symbol_table=symbol_table)
    elif fonction.name in symbol_table:
        gen_appel_fonction_user(fonction, symbol_table)
    else:
        print("fonction inconnue", fonction.name)
        exit(0)


def gen_appel_fonction_user(fonction: arbre_abstrait.AppelFonction,
                            symbol_table: SymbolTable):
    """
    Affiche le code nasm correspondant à l'appel d'une fonction utilisateur
    """
    # on empile les arguments de la fonction
    for arg in fonction.args:
        gen_expression(arg, symbol_table)

    # on appelle la fonction
    nasm_instruction("call", f"_{fonction.name.valeur}", "", "", "")

    # on empile la valeur de retour de la fonction
    nasm_instruction("push", "eax", "", "", "")


def gen_function(fonction: arbre_abstrait.Function, symbol_table: SymbolTable):
    """
    Affiche le code nasm correspondant à la définition d'une fonction
    """
    printift(fonction)

    global current_function
    current_function = fonction

    printifm(f"_{fonction.name.valeur}:")
    gen_listeInstructions(fonction.body.instructions, symbol_table)

    current_function = None


def gen_expression(expression: arbre_abstrait.AST, symbol_table: SymbolTable):
    """
    Affiche le code nasm pour calculer et empiler la valeur d'une expression
    """
    if type(expression) == arbre_abstrait.Operation:
        # on calcule et empile la valeur de l'opération
        gen_operation(expression, symbol_table)
    elif type(expression) == arbre_abstrait.OperationUnaire:
        gen_operation_unaire(expression, symbol_table)
    elif type(expression) == arbre_abstrait.Entier or type(
            expression) == arbre_abstrait.Booleen:
        # on met sur la pile la valeur entière
        nasm_instruction("push", str(int(expression.valeur)), "", "", "")
    elif type(expression) == arbre_abstrait.AppelFonction:
        gen_appel_fonction(expression, symbol_table)
    else:
        print("type d'expression inconnu", type(expression))
        exit(0)


def gen_operation(operation: arbre_abstrait.Operation,
                  symbol_table: SymbolTable):
    """
    Affiche le code nasm pour calculer l'opération et la mettre en haut de la pile
    """

    op = operation.op

    gen_expression(operation.lhs,
                   symbol_table)  # on calcule et empile la valeur de exp1
    gen_expression(operation.rhs,
                   symbol_table)  # on calcule et empile la valeur de exp2

    nasm_instruction("pop", "ebx", "", "",
                     "dépile la seconde operande dans ebx")
    nasm_instruction("pop", "eax", "", "",
                     "dépile la permière operande dans eax")

    code = {
        "+": "add",
        "*": "imul",
        "-": "sub",
        "/": "div",
        "%": "mod",
        "et": "and",
        "ou": "or",
    }
    # Un dictionnaire qui associe à chaque opérateur sa fonction nasm
    # Voir: https://www.bencode.net/blob/nasmcheatsheet.pdf
    if op in ['+', '-']:
        check_type(operation, arbre_abstrait.Type.ENTIER)

        nasm_instruction(
            code[op], "eax", "ebx", "", "effectue l'opération eax" + op +
            "ebx et met le résultat dans eax")
    elif op == '*':
        check_type(operation, arbre_abstrait.Type.ENTIER)

        nasm_instruction(
            code[op], "ebx", "", "", "effectue l'opération eax" + op +
            "ebx et met le résultat dans eax")
    elif op == '/':
        check_type(operation, arbre_abstrait.Type.ENTIER)

        nasm_instruction("mov", "edx", "0", "", "met edx à 0")

        nasm_instruction(
            code[op], "ebx", "", "", "effectue l'opération edx:eax" + op +
            "ebx et met le résultat dans eax")
    elif op == '%':
        check_type(operation, arbre_abstrait.Type.ENTIER)

        nasm_instruction("mov", "edx", "0", "", "met edx à 0")
        nasm_instruction(
            "idiv", "ebx", "", "", "effectue l'opération eax" + op +
            "ebx et met le résultat dans edx")
    elif op in ["ou", "et"]:
        check_type(operation, arbre_abstrait.Type.BOOLEEN)

        nasm_instruction(
            code[op], "eax", "ebx", "", "effectue l'opération eax" + op +
            "ebx et met le résultat dans eax")

    elif op in ["<", "<=", ">", ">=", "==", "!="]:
        gen_comparison(operation)

    else:
        print("type d'opération inconnu", op)
        exit(0)

    target = "edx" if op == "%" else "eax"

    nasm_instruction("push", target, "", "", "empile le résultat")


def gen_operation_unaire(operation: arbre_abstrait.OperationUnaire,
                         symbol_table: SymbolTable):
    op = operation.op

    gen_expression(operation.exp,
                   symbol_table)  # on calcule et empile la valeur de exp
    nasm_instruction("pop", "eax", "", "",
                     "dépile la première operande dans eax")

    if op == '-':
        check_type(operation, arbre_abstrait.Type.ENTIER)
        nasm_instruction("neg", "eax", "", "", "effectue l'opération -eax")
    elif op == "non":
        check_type(operation, arbre_abstrait.Type.BOOLEEN)
        nasm_instruction("xor", "eax", "1", "", "effectue l'opération not eax")
    else:
        print("type d'opération unaire inconnu")
        exit(0)

    nasm_instruction("push", "eax", "", "", "empile le résultat")


def gen_comparison(expr: arbre_abstrait.Operation):
    """
    Génère le code pour une comparaison
    """

    check_type(expr, arbre_abstrait.Type.ENTIER)

    # On effectue la comparaison
    nasm_instruction("cmp", "eax", "ebx", "", "compare eax et ebx")
    if expr.op == '<':
        nasm_instruction("setl", "al", "", "", "met al à 1 si eax < ebx")
    elif expr.op == '<=':
        nasm_instruction("setle", "al", "", "", "met al à 1 si eax <= ebx")
    elif expr.op == '>':
        nasm_instruction("setg", "al", "", "", "met al à 1 si eax > ebx")
    elif expr.op == '>=':
        nasm_instruction("setge", "al", "", "", "met al à 1 si eax >= ebx")
    elif expr.op == '==':
        nasm_instruction("sete", "al", "", "", "met al à 1 si eax == ebx")
    elif expr.op == '!=':
        nasm_instruction("setne", "al", "", "", "met al à 1 si eax != ebx")
    else:
        print("type de comparaison inconnu")
        exit(0)


def main():
    global afficher_nasm
    global afficher_tableSymboles

    afficher_nasm = True
    if len(sys.argv) < 3 or sys.argv[1] not in ["-nasm", "-table"]:
        print(
            "usage: python3 generation_code.py -nasm|-table NOM_FICHIER_SOURCE.flo"
        )
        exit(0)
    if sys.argv[1] == "-nasm":
        afficher_nasm = True
    if sys.argv[1] == "-table":
        afficher_tableSymboles = True

    with open(sys.argv[2], "r") as f:
        data = f.read()
        try:
            arbre = analyse_syntaxique(data)
            gen_entrypoint(arbre)
        except EOFError:
            exit()


if __name__ == "__main__":
    main()
