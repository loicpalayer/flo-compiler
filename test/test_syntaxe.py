import syrupy
from src.analyse_syntaxique import analyse_syntaxique
import json


def test_syntaxe(snapshot):
    input = """
    ecrire(- (a + b - c) * d / e % f);
    fonction1(a et non(b ou c));
    bonsoir(a == b != c < d <= e > f >= g);
    """

    out = analyse_syntaxique(input)
    actual = json.dumps(out, indent=2)

    assert snapshot == actual
