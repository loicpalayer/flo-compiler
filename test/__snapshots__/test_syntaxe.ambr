# serializer version: 1
# name: test_declare_assign
  '''
  {
    "instructions": [
      {
        "name": {
          "identifiant": "a"
        },
        "type": "ENTIER",
        "value": null
      },
      {
        "name": {
          "identifiant": "a"
        },
        "value": {
          "entier": 1
        }
      },
      {
        "name": {
          "identifiant": "b"
        },
        "type": "BOOLEEN",
        "value": {
          "identifiant": "vrai"
        }
      }
    ]
  }
  '''
# ---
# name: test_function_call
  '''
  {
    "instructions": [
      {
        "function_name": "fonction1",
        "args": [
          {
            "identifiant": "a"
          },
          {
            "identifiant": "b"
          },
          {
            "identifiant": "c"
          }
        ]
      }
    ]
  }
  '''
# ---
# name: test_parse_if
  '''
  {
    "instructions": [
      {
        "condition": {
          "identifiant": "a"
        },
        "body": {
          "instructions": [
            {
              "function_name": "ecrire",
              "args": [
                {
                  "identifiant": "a"
                }
              ]
            }
          ]
        },
        "orelse": {
          "instructions": [
            {
              "condition": {
                "identifiant": "b"
              },
              "body": {
                "instructions": [
                  {
                    "function_name": "ecrire",
                    "args": [
                      {
                        "identifiant": "b"
                      }
                    ]
                  }
                ]
              },
              "orelse": {
                "instructions": [
                  {
                    "condition": {
                      "identifiant": "c"
                    },
                    "body": {
                      "instructions": [
                        {
                          "function_name": "ecrire",
                          "args": [
                            {
                              "identifiant": "c"
                            }
                          ]
                        },
                        {
                          "function_name": "ecrire",
                          "args": [
                            {
                              "identifiant": "d"
                            }
                          ]
                        }
                      ]
                    },
                    "orelse": {
                      "instructions": []
                    }
                  }
                ]
              }
            },
            {
              "instructions": [
                {
                  "function_name": "ecrire",
                  "args": [
                    {
                      "identifiant": "c"
                    }
                  ]
                }
              ]
            }
          ]
        }
      }
    ]
  }
  '''
# ---
# name: test_parse_while
  '''
  {
    "instructions": [
      {
        "condition": {
          "identifiant": "a"
        },
        "body": {
          "instructions": [
            {
              "function_name": "ecrire",
              "args": [
                {
                  "identifiant": "a"
                }
              ]
            }
          ]
        }
      }
    ]
  }
  '''
# ---
# name: test_return
  '''
  {
    "instructions": [
      {
        "return_value": {
          "entier": 1
        }
      }
    ]
  }
  '''
# ---
# name: test_syntaxe
  '''
  {
    "instructions": [
      {
        "function_name": "ecrire",
        "args": [
          {
            "op": "%",
            "lhs": {
              "op": "/",
              "lhs": {
                "op": "*",
                "lhs": {
                  "op": "-",
                  "exp": {
                    "op": "-",
                    "lhs": {
                      "op": "+",
                      "lhs": {
                        "identifiant": "a"
                      },
                      "rhs": {
                        "identifiant": "b"
                      }
                    },
                    "rhs": {
                      "identifiant": "c"
                    }
                  }
                },
                "rhs": {
                  "identifiant": "d"
                }
              },
              "rhs": {
                "identifiant": "e"
              }
            },
            "rhs": {
              "identifiant": "f"
            }
          }
        ]
      },
      {
        "function_name": "fonction1",
        "args": [
          {
            "op": "et",
            "lhs": {
              "identifiant": "a"
            },
            "rhs": {
              "op": "non",
              "exp": {
                "op": "ou",
                "lhs": {
                  "identifiant": "b"
                },
                "rhs": {
                  "identifiant": "c"
                }
              }
            }
          }
        ]
      },
      {
        "function_name": "bonsoir",
        "args": [
          {
            "op": "==",
            "lhs": {
              "identifiant": "a"
            },
            "rhs": {
              "op": ">=",
              "lhs": {
                "op": ">",
                "lhs": {
                  "op": "<=",
                  "lhs": {
                    "op": "<",
                    "lhs": {
                      "op": "!=",
                      "lhs": {
                        "identifiant": "b"
                      },
                      "rhs": {
                        "identifiant": "c"
                      }
                    },
                    "rhs": {
                      "identifiant": "d"
                    }
                  },
                  "rhs": {
                    "identifiant": "e"
                  }
                },
                "rhs": {
                  "identifiant": "f"
                }
              },
              "rhs": {
                "identifiant": "g"
              }
            }
          }
        ]
      }
    ]
  }
  '''
# ---
